src/tracks/SampleTrack.cpp

```diff
@@ -572,58 +574,119 @@ SampleTrack::~SampleTrack()
 
 
 
-bool SampleTrack::play( const MidiTime & _start, const fpp_t _frames,
-					const f_cnt_t _offset, int _tco_num )
+bool SampleTrack::play( const MidiTime & _start, const fpp_t /*_frames*/,
+						const f_cnt_t _offset, int _tco_num )
 {
 	m_audioPort.effects()->startRunning();
-	bool played_a_note = false;	// will be return variable
+
+	//will be return variable
+	bool played_a_note = false;
 
 	tcoVector tcos;
-	::BBTrack * bb_track = NULL;
+
+	::BBTrack * bb_track = nullptr;
+	MidiTime bbEndPos = 0;
+
+	/********************************************************************
+	* In this whole block we find our BB Track TCO in the Songeditor
+	* and so we can calculate the length of it.
+	* If _tco_num is equal or bigger than 0, we know we have a
+	* Sampletrack within a BB-Container
+	********************************************************************/
+
 	if( _tco_num >= 0 )
 	{
-		if( _start != 0 )
+		//repeat playing when end of BB-Container is reached
+		tick_t lengthOfBB = Engine::getBBTrackContainer()->lengthOfBB( _tco_num ) * MidiTime::ticksPerTact();
+		if( _start % lengthOfBB == 0 )
 		{
-			return false;
+			updateTcos();
 		}
-		tcos.push_back( getTCO( _tco_num ) );
-		if (trackContainer() == (TrackContainer*)Engine::getBBTrackContainer())
+
+		//the BB-Track in Songeditor
+		bb_track = BBTrack::findBBTrack( _tco_num );
+		//Vector of BB Tcos in Songeditor. It will be filled later
+		tcoVector bbtcos;
+		//Do we play in BB-Editor or in Songeditor?
+		if( Engine::getSong()->playMode() != Song::Mode_PlayBB )
 		{
-			bb_track = BBTrack::findBBTrack( _tco_num );
+			MidiTime currentSongPos = Engine::getSong()->getPlayPos( Song::Mode_PlaySong ).getTicks();
+			bb_track->getTCOsInRange( bbtcos, currentSongPos, currentSongPos );
+			//take care of overlapping BB-TCOs
+			if( bbtcos.size() != m_NumOfOverlappingBB )
+			{
+				m_NumOfOverlappingBB = bbtcos.size();
+				updateTcos();
+			}
+			//play the TCO which comes at last
+			auto startpos = 0;
+			for( const auto it : bbtcos )
+			{
+				if( !it->isMuted() && it->startPosition() > startpos )
+				{
+					startpos = it->startPosition();
+					//how many times the bb is looped?
+					int loopCountBB = ( currentSongPos - startpos ) / lengthOfBB;
+					bbEndPos = it->length() - loopCountBB * lengthOfBB;
+				}
+			}
 		}
 	}
-	else
+
+	/********************************************************************
+	 * Now we iterate over all TCOs in the Sampletrack
+	 * If the sampletrack is within a BB-Container we have only one TCO
```

Comment:
Thank you, I look into it tomorrow.

Reply:
Confirmed that when there are two or more BB tracks and you press play inside one of these BB tracks then samples from all BB tracks play at the same time.

Reply:
Multiple samples in BB Tracks is wonky already in stable-1.2 . It's not obvious how you are supposed to use them (not to me anyway). I think it would be good with a reference project based on the lmms samples for testing this.